statsmodels/sandbox/gam_gsoc2015/gam.py

```diff
@@ -0,0 +1,381 @@
+            raise ValueError('all the input values should be list of the same length')
+
+        # the total number of columns in der2 i.e. the len of the params vector
+        self.k_columns = np.sum(d2.shape[1] for d2 in der2)
+
+        # the number of variables in the GAM model
+        self.n_variables = len(cov_der2)
+
+        # if wts and alpha are not a list then each function has the same penalty
+        # TODO: Review this
+        self.alphas = alphas
+        self.wts = wts
+
+        n_samples = der2[0].shape[0]
+        self.mask = [np.array([False]*self.k_columns)
+                     for _ in range(self.n_variables)]
+        param_count = 0
+        for i, d2 in enumerate(der2):
+            n, dim_base = d2.shape
+            # check that all the basis have the same number of samples
+            assert(n_samples == n)
+            # the mask[i] contains a vector of length k_columns. The index
+            # corresponding to the i-th input variable are set to True.
+            self.mask[i][param_count: param_count + dim_base] = True
+            param_count += dim_base
+
+        self.gp = []
+        for i in range(self.n_variables):
+            gp = GamPenalty(wts=self.wts[i], alpha=self.alphas[i],
+                            cov_der2=cov_der2[i], der2=der2[i])
+            self.gp.append(gp)
+
+        return
+
+    def func(self, params):
+        cost = 0
+        for i in range(self.n_variables):
+            params_i = params[self.mask[i]]
+            cost += self.gp[i].func(params_i)
+
+        return cost
+
+    def grad(self, params):
+        grad = []
+        for i in range(self.n_variables):
+            params_i = params[self.mask[i]]
+            grad.append(self.gp[i].grad(params_i))
+
+        return np.concatenate(grad)
+
+    def deriv2(self, params):
+        deriv2 = np.empty(shape=(0,0))
+        for i in range(self.n_variables):
+            params_i = params[self.mask[i]]
+            deriv2 = block_diag(deriv2, self.gp[i].deriv2(params_i))
+
+        return deriv2
+
+
+class LogitGam(PenalizedMixin, Logit):
+    pass
+
+
+class GLMGAMResults(GLMResults):
+
+    def plot_predict(self, x_values=None, smooth_basis=None):
+        """just to try a method in overridden Results class
+        """
+        import matplotlib.pyplot as plt
+        # TODO: This function will be available when we will have the self.model.x variable
+        # if x_values is None:
+        #     plt.plot(self.model.x, self.model.endog, '.')
+        #     plt.plot(self.model.x, self.predict())
+        # else:
+        plt.plot(x_values, self.predict(smooth_basis))
+
+    def significance_test(self, basis=None):
+        v = basis.dot(self.normalized_cov_params).dot(basis.T)
+        p_inv_v = pinv(v)
+        hat_y = self.predict(basis)
+        tr = hat_y.T.dot(p_inv_v).dot(hat_y)
+
+        lin_pred = np.dot(basis, self.params)
+        mu = self.family.fitted(lin_pred)
```

Comment:
`linpred`, `mu` need to be based on all exog (and offset exposure) outside of the linear gaussian model, or if link is not the identity. (that's because of the nonlinear transformation by the link)

(It shouldn't affect the results below if that's for linear gaussian default case.)